pkg/mapper/mapper_cache.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2019 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mapper
+
+import (
+	"github.com/hashicorp/golang-lru"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	cacheLength = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Name: "statsd_exporter_cache_length",
+			Help: "The count of unique metrics currently cached.",
+		},
+	)
+)
+
+type MetricMapperCacheResult struct {
+	Mapping *MetricMapping
+	Matched bool
+	Labels  prometheus.Labels
+}
+
+type MetricMapperCache struct {
+	cache *lru.Cache
+}
+
+func NewMetricMapperCache(size int) (*MetricMapperCache, error) {
+	cacheLength.Set(0)
```

Comment:
is the cacheLength global or per cache? It seems that everything is set up to accommodate multiple metric mapper caches, except for the cacheLength gauge.

I would propose having a prometheus.Gauge passed in by the caller instead?

Reply:
It's global, I don't expect to have multiple metric mapper caches, I guess there's nothing stopping you, but I'm not 100% sure what you gain, since the cache implementation is thread safe. I guess decreased lock contention, but in my production test of this with a more threaded backend focusing into one instance of the cache (we've had it running with a cache ~1/2 the size of the incoming metrics to force a high churn), lock contention has not been a problem. I'm not against doing this, but I kind of feel like it's something we can swap to later if we decide it's a path we wanna head down?

Reply:
My concern is less about what we need right now, and more about the package interface being confusing – however I just realized that we already have a metric like that, and you're right that it's not something we need to fix right now, so I'm okay with it as it is.